llama-index-core/llama_index/core/indices/property_graph/sub_retrievers/custom.py

```diff
@@ -0,0 +1,49 @@
+from abc import abstractmethod
+from typing import Any, List
+
+from llama_index.core.graph_stores.types import PropertyGraphStore
+from llama_index.core.indices.property_graph.sub_retrievers.base import BasePGRetriever
+from llama_index.core.schema import NodeWithScore, QueryBundle, TextNode
+
+
+class CustomPGRetriever(BasePGRetriever):
+    """A retriever meant to be easily subclassed to implement custom retrieval logic.
+
+    The user only has to implement:
+    - `init` to initialize the retriever and assign any necessary attributes.
+    - `custom_retrieve` to implement the custom retrieval logic.
+    - `aretrieve_custom` (optional) to implement asynchronous retrieval logic.
+
+    Args:
+        graph_store (PropertyGraphStore):
+            The graph store to retrieve data from.
+    """
+
+    def __init__(
+        self,
+        graph_store: PropertyGraphStore,
+        **kwargs: Any,
+    ) -> None:
+        self.init(**kwargs)
+        super().__init__(graph_store=graph_store, include_text=False, **kwargs)
+
+    @abstractmethod
+    def init(self, **kwargs: Any):
+        ...
+
+    @abstractmethod
+    def custom_retrieve(self, query_str: str) -> str:
```

Comment:
why is the return type a single string? wouldn't it make more sense to just let the user return a list of Nodes? If you don't want to expose NodeWithScore you can always just have the user expose the text node (and attach NodeWithScore as optional)

Reply:
Yea just though a string would be easiest to write when subclassing 😅 can make it a text node

Reply:
asking because i'm defining a custom retriever right now :)